Metrics APIs

Mazen-Nayef opened the pull request:
1- Make Create, Delete, Put, Get/ID, Get All Metrics APIs in the file metric in folder api .
2- Created File exceptions & handling_exception files in api folder to handle all the ExceptionCustom. 
3- Changed in the base file in crud folder replaced all the HTTPException with ExceptionCustom.
4- Added Exception handler in main to handle the error message of datatype of pydantic and change its shape.

Mazen-Nayef committed:
1- Added New Post, Get/ID API for Metrics.
2- Made a Custom Error Message for the datatype validation of pydantic
3- Changed the schema of Metric to get List of strings in the area field and be stringified and parsed back to a list when it is called back.
4- getByName function added in the CRUDMetrics to get by the name of the Metric.

Mazen-Nayef committed:
1- Added a validator for the area field to make sure it is not empty list.
2- Made delete API for metric to delete any metric by using its own ID

Mazen-Nayef committed:
1- Added put api metric but still need some work to be finsished

Mazen-Nayef committed:
1- Added PUT API for metrics
2- Changed in metric Schema as a make the validator for the create schema only

Mazen-Nayef committed:
1- Handled all the exceptions using the ExceptionCustom
2- Created file exceptions and handling_exception
3- Added add_exception_handler in main.py
4- Changed in the base.py where i replaced the HTTPExceptions with ExceptionCustom

Mazen-Nayef committed:
Removed unused parts of the code

Mazen-Nayef committed:
Resolved all the conflicts between my branch and the main-backend

Passant-Abdelgalil commented on app/api/metrics.py:
```diff
@@ -1,17 +1,30 @@
 router = APIRouter()
+app = FastAPI()
 
+def format_code(name):
```
Extract this function to a file called index.py or something inside a new folder called utils and import it
Make sure to update the microservice.py api file to use the function from the utils file

Passant-Abdelgalil commented on app/api/metrics.py:
```diff
@@ -51,3 +63,61 @@ def getScoreCards(
+    # change ' ' with '-'
+    metricObj.code = format_code(metric.name)
+    # Stringfying the list of areas to be saved as string 
+    metricObj.area = json.dumps(metric.area)
```
check if area is not None before calling json.dumps
in general, json.dumps will throw exception in case something went wrong, make sure to handle this exception and throw an ExceptionCustom with the right status_code and a descriptive error message.

Passant-Abdelgalil commented on app/api/metrics.py:
```diff
@@ -51,3 +63,61 @@ def getScoreCards(
+    # Stringfying the list of areas to be saved as string 
+    metricObj.area = json.dumps(metric.area)
+    # Handling the datatype of the field type to be integer or boolean 
+    if (metricObj.type != "integer" and metricObj.type != "boolean"):
```
this will not extend smoothly when you want to add a new metric type.
suggestion: define a list of acceptable types, and update the if condition to check if the metricObj.type is within the defined list or not

Passant-Abdelgalil commented on app/api/metrics.py:
```diff
@@ -51,3 +63,61 @@ def getScoreCards(
+    if (metricObj.type != "integer" and metricObj.type != "boolean"):
+        raise ExceptionCustom(status_code=422, detail= "type must be integer or boolean")
+    metricCrud.create(metricObj)
+    raise ExceptionCustom(status_code=200, detail="Success in creating metric")
```
rename HTTPResponseCustom or something

Passant-Abdelgalil commented on app/api/metrics.py:
```diff
@@ -51,3 +63,61 @@ def getScoreCards(
+@router.get("/{metricID}", response_model=schemas.List[schemas.Metric])
+def getMetric(metricID: int, metricCrud: crud.CRUDMetric = Depends(dependencies.getMetricsCrud)) -> Any:
+    metric = metricCrud.get(metricID)
+    metric.area = json.loads(metric.area)
```
json.loads may throw exception if database contains invalid data for metric.area for any reason.
make sure to handle exceptions and in case of any exception for the metric.are, return the metric data but with empty area field.

Passant-Abdelgalil commented on app/api/metrics.py:
```diff
@@ -51,3 +63,61 @@ def getScoreCards(
+def getMetric(metricID: int, metricCrud: crud.CRUDMetric = Depends(dependencies.getMetricsCrud)) -> Any:
+    metric = metricCrud.get(metricID)
+    metric.area = json.loads(metric.area)
+    metricOBJ = jsonable_encoder(metric)
```
praise: thank you for paying attention to this!

Passant-Abdelgalil commented on app/api/metrics.py:
```diff
@@ -51,3 +63,61 @@ def getScoreCards(
+    metricObj = metricInput
+    if (metricInput.name):
+        metricObj.name = metricInput.name
+        metricObj.code = format_code(metricInput.name)
```
praise: good catch and nice work to put a constraint on the DB column!
make sure to handle the DB exception and rethrow an ExceptionCustom with a descriptive error message.

Passant-Abdelgalil commented on app/crud/base.py:
```diff
@@ -34,7 +34,7 @@ def create(self, obj: CreateSchemaType) -> ModelType:
             self.db_session.rollback()
             if "duplicate key" in str(e):
-                raise ExceptionCustom(status_code=409, detail="Conflict Error")
+                raise ExceptionCustom(status_code=409, detail="Duplicated Key, Name field is Duplicated")
```
praise: thank you for paying attention to this Exception!
suggestion: extract the information of the field from the error object itself, don't hardcode it

Passant-Abdelgalil commented on app/main.py:
```diff
@@ -22,6 +26,24 @@ def create_app() -> FastAPI:
 
+    
+    # Edit on the exceptiion handler of data type validation pydantic to send custom error messages
+    @_app.exception_handler(RequestValidationError)
```
praise: nice work and great job for customizing the Exception!
suggestion: extract the handling logic into a function and move it to handling_exception file and import it from their

Passant-Abdelgalil commented on app/schemas/metric.py:
```diff
@@ -1,21 +1,32 @@
     name: str
     description: str
 
+    @validator('area', pre=True, always=True)
```
praise: nice work for search for validators!
suggestion: treat empty list as if it's was passed as None inside the api route handler and don't throw a ValueError

Mazen-Nayef committed:
1- Removed the exception handler of the data types from the main.py and put it in file handling_exception.py
2- made a metric_type list to be easily expanded and edited in the future
3- removed the format_code function and put it in utils.base file to be imported anywhere we need it
4- checks now on the area of the posted metric if it has empty string it returns error and if it is not passed i put it an empty list
5- channged the name of ExceptionCustom => HTTPResponseCustomized
6- checks now on the area in the single get of the metric we try first to loads it if we can not we return an empty list
7- checks now on the edit if the code is repeated or not and raise exception if it is repeated & in the same edit we checks if we can dump the area if we can not we put it as an empty list
8- i made a small change of the list CRUD found in the base.py as i now sort it by the ID ascendingly
9- removed the validator of the empty list of the area of metric
